synapse/handlers/oidc_handler.py

```diff
@@ -0,0 +1,840 @@
+                            'alg': 'RS256',
+                            'use': 'sig',
+                            'e': 'XXXX',
+                            'n': 'XXXX',
+                        }
+                    ]
+                }
+        """
+        if self._uses_userinfo:
+            # We're not using jwt signing, return an empty jwk set
+            return {"keys": []}
+
+        # First check if the JWKS are loaded in the provider metadata.
+        # It can happen either if the provider gives its JWKS in the discovery
+        # document directly or if it was already loaded once.
+        metadata = await self.load_metadata()
+        jwk_set = metadata.get("jwks")
+        if jwk_set is not None and not force:
+            return jwk_set
+
+        # Loading the JWKS using the `jwks_uri` metadata
+        uri = metadata.get("jwks_uri")
+        if not uri:
+            raise RuntimeError('Missing "jwks_uri" in metadata')
+
+        jwk_set = await self._http_client.get_json(uri)
+
+        # Caching the JWKS in the provider's metadata
+        self._provider_metadata["jwks"] = jwk_set
+        return jwk_set
+
+    async def _exchange_code(self, code: str) -> Token:
+        """Exchange an authorization code for a token.
+
+        This calls the ``token_endpoint`` with the authorization code we
+        received in the callback to exchange it for a token. The call uses the
+        ``ClientAuth`` to authenticate with the client with its ID and secret.
+
+        Args:
+            code: The autorization code we got from the callback.
+
+        Returns:
+            dict: contains various tokens.
+
+            May look like this::
+
+                {
+                    'token_type': 'bearer',
+                    'access_token': 'abcdef',
+                    'expires_in': 3599,
+                    'id_token': 'ghijkl',
+                    'refresh_token': 'mnopqr',
+                }
+
+        Raises:
+            OidcError: when the ``token_endpoint`` returned an error.
+        """
+        metadata = await self.load_metadata()
+        token_endpoint = metadata.get("token_endpoint")
+        headers = {
+            "Content-Type": "application/x-www-form-urlencoded",
+            "User-Agent": self._http_client.user_agent,
+            "Accept": "application/json",
+        }
+
+        args = {
+            "grant_type": "authorization_code",
+            "code": code,
+            "redirect_uri": self._callback_url,
+        }
+        body = urlencode(args, True)
+
+        # Fill the body/headers with credentials
+        uri, headers, body = self._client_auth.prepare(
+            method="POST", uri=token_endpoint, headers=headers, body=body
+        )
+        headers = {k: [v] for (k, v) in headers.items()}
+
+        # Do the actual request
+        # We're not using the SimpleHttpClient util methods as we don't want to
+        # check the HTTP status code and we do the body encoding ourself.
```

Comment:
We don't check the status code because the important thing is whether there's an `"error"` key in the response?

Is this true for like a 500 response as well or are we just expecting 200 or 4xx?

Reply:
Reading through the OAuth2 RFC ([RFC6749 sec 5.2](https://tools.ietf.org/html/rfc6749#section-5.2)), it seems like errors should be 400 or 401 codes, while success should always be 200. Grepping through Keycloak's code, it seems that they are also using 403 codes.

I think in all cases, we should check for 5xx errors and nicely show/log those errors. For other cases, either: 

 - check the status code, and if it's a 2xx, assume there is no error and don't check for the `error*` fields. In case of a misbehaving provider, it will throw later in the code, resulting to a `M_UNKNOWN` error (which should never happen).
 - check for the `error` field, regardless of the status code
 - check both, and throw if the provider is misbehaving (e.g. a 200 with a "error" field or a 4xx without one)

Reply:
I think we probably want to reject all 500 errors, it is likely that those won't even be a JSON response (although I guess an error would be thrown earlier in the process in that case), so maybe this is OK.

Reply:
That's done. I'm checking that http status code & error field matching. In case of a 5xx, I *try* to parse the JSON and extract the "error" field, but if it fails, it returns a generic server error.

This part is heavily commented and all branches are tested, so you should not have trouble figuring what it is doing